spec/services/waste_exemptions_engine/registration_completion_service_spec.rb

```diff
@@ -10,6 +10,55 @@ module WasteExemptionsEngine
     let(:registration_completion_service) { RegistrationCompletionService.new(new_registration) }
 
     describe "#complete" do
+      it "is idempotent" do
+        expect(WasteExemptionsEngine::Registration.count).to eq(0)
+        RegistrationCompletionService.new(new_registration).complete
+
+        expect { RegistrationCompletionService.new(new_registration).complete }.to raise_error
+
+        expect(WasteExemptionsEngine::Registration.count).to eq(1)
+      end
+
+      context "when a race condition calls the service twice" do
+        # rubocop:disable Lint/HandleExceptions
+        it "generates only one record and fail to execute subsequent calls" do
+          expect(WasteExemptionsEngine::Registration.count).to eq(0)
+          expect(ActiveRecord::Base.connection.pool.size).to be > 3
+
+          should_wait = true
+          concurrency_level = 4
+
+          begin
+            threads = Array.new(concurrency_level) do
+              Thread.new do
+                while should_wait
+                  # Do nothing, just wait
+                end
+
+                begin
+                  registration_completion_service.complete
+                rescue StandardError
+                end
+              end
+            end
+          ensure
+            ActiveRecord::Base.connection_pool.disconnect!
+          end
+
+          should_wait = false
+          threads.each(&:join)
+
+          expect(WasteExemptionsEngine::Registration.count).to eq(1)
+
+          # Clean up after the threads have executed
```

Comment:
DatabaseCleaner should be configured to clean up around each example, so I'm not sure we need to do this ourselves - was this because of a specific issue?

Reply:
This is because we use a transaction strategy, and Database cleaner can't do transactions with Threads, as they happen outside of the rspec process

Reply:
@cintamani Got it 👍 I knew there must have been a reason!